rulesets_command.php

```diff
@@ -0,0 +1,110 @@
+        $repo = $module['repo'];
+
+        // Fetch existing rulesets
+        $rulesets = github_api("https://api.github.com/repos/$account/$repo/rulesets");
+        $branchRulesetID = 0;
+        $tagRulesetID = 0;
+        foreach ($rulesets as $ruleset) {
+            $id = $ruleset['id'];
+            $name = $ruleset['name'];
+            if ($name === BRANCH_RULESET_NAME) {
+                $branchRulesetID = $id;
+            }
+            if ($name === TAG_RULESET_NAME) {
+                $tagRulesetID = $id;
+            }
+        }
+
+        // Get any additional branches to add
+        // Assumption is that if the default branch is main/master, then the repo uses
+        // a non-numeric style branching system (e.g. main, master) and that needs to be protected
+        // [0-9]* branch protection will still be applied, on the chance that the repo is converted
+        // to uses a numeric style branch system in the future and we would want branch protection
+        // to start immediately on the new branches
+        $additionalBranchConditions = [];
+        $defaultBranch = github_api("https://api.github.com/repos/$account/$repo")['default_branch'];
+        if (in_array($defaultBranch, ['main', 'master'])) {
+            $additionalBranchConditions[] = "refs/heads/$defaultBranch";
+        }
+
+        // Create rulesets
+        // Note: This will read from the "rulesets" directory
+        // In each of those json rulesets there is "bypass_actors"."actor_id" = 5
+        // This translates to the "Repository admin" role
+        // It has been confirmed that the github-action user is able to bypass the ruleset as
+        // it has the "Organisation admin" role which is one level above the "Repository admin" role
+        $branchRuleset = create_ruleset('branch', $additionalBranchConditions);
+        $tagRuleset = create_ruleset('tag');
+
+        $apiCalls = [];
+
+        // Create new rulesets
+        if (!$branchRulesetID) {
+            $url = "https://api.github.com/repos/$account/$repo/rulesets";
+            $apiCalls[] = [$url, $branchRuleset, 'POST'];
+        }
+        if (!$tagRulesetID) {
+            $url = "https://api.github.com/repos/$account/$repo/rulesets";
+            $apiCalls[] = [$url, $tagRuleset, 'POST'];
+        }
+
+        // Update existing rulesets
+        // Don't bother to check if the ruleset is already correct
+        // This is a very quick update so no need to optimise this
+        if ($branchRulesetID) {
+            $url = "https://api.github.com/repos/$account/$repo/rulesets/$branchRulesetID";
+            $apiCalls[] = [$url, $branchRuleset, 'PUT'];
+        }
+        if ($tagRulesetID) {
+            $url = "https://api.github.com/repos/$account/$repo/rulesets/$tagRulesetID";
+            $apiCalls[] = [$url, $tagRuleset, 'PUT'];
+        }
+
+        if ($input->getOption('dry-run')) {
+            info('Not updating rulesets on GitHub because --dry-run option is set');
```

Comment:
May be worth dumping `$apiCalls` either to a file or to the terminal here so we can see in a dry-run if the calls look like we expect them to.

Reply:
Done